### PR TITLE
IndexTemplateMetaData should only store a single mapping

### PR DESCRIPTION
**Pull request by romseygeek**
Index templates can only store a single mapping, but `getMappings()` returns a map
of mapping type to mappings.  This commit changes the signature of `getMappings()` to
just return the mappings.

In addition, index templates stored in the cluster state will now always convert their type
to `_doc`, and templates added via `PutIndexTemplateRequest` will also convert their
type.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java [hunk 1]
```diff
@@ -77,14 +79,13 @@
...
 
-    // the mapping source should always include the type as top level
     private final ImmutableOpenMap<String, CompressedXContent> mappings;
 
     private final ImmutableOpenMap<String, AliasMetaData> aliases;
 
     public IndexTemplateMetaData(String name, int order, Integer version,
                                  List<String> patterns, Settings settings,
-                                 ImmutableOpenMap<String, CompressedXContent> mappings,
+                                 CompressedXContent mappings,
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java]: It is not very clear if we expect `CompressedXContent  mappings` to contain the type or not?  From line 98, it looks like it should not contain type, but then how `IndexTemplateMetaData` is built  `fromXContent` and ` builder` it looks like `mappings` already contain a type? 
Or we can expect both cases?

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java [hunk 2]
```diff
@@ -281,13 +302,13 @@ public Builder settings(Settings settings) {
             return this;
         }
 
-        public Builder putMapping(String mappingType, CompressedXContent mappingSource) {
-            mappings.put(mappingType, mappingSource);
+        public Builder setMapping(CompressedXContent mappingSource) {
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java]: here it is not very clear if we expect `mappingSource` to contain a type or not.  I saw examples of both

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java [hunk 3]
```diff
@@ -444,10 +467,10 @@ public static IndexTemplateMetaData fromXContent(XContentParser parser, String t
                             if (token == XContentParser.Token.FIELD_NAME) {
                                 currentFieldName = parser.currentName();
                             } else if (token == XContentParser.Token.START_OBJECT) {
-                                String mappingType = currentFieldName;
+                                String mappingType = MapperService.SINGLE_MAPPING_NAME;
                                 Map<String, Object> mappingSource =
                                     MapBuilder.<String, Object>newMapBuilder().put(mappingType, parser.mapOrdered()).map();
-                                builder.putMapping(mappingType, Strings.toString(XContentFactory.jsonBuilder().map(mappingSource)));
+                                builder.setMapping(Strings.toString(XContentFactory.jsonBuilder().map(mappingSource)));
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java]: here looks like `setMapping` expects mapping with a type

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java [hunk 4]
```diff
@@ -94,7 +95,8 @@ public IndexTemplateMetaData(String name, int order, Integer version,
         this.version = version;
         this.patterns = patterns;
         this.settings = settings;
-        this.mappings = mappings;
+        this.mappings = ImmutableOpenMap.<String, CompressedXContent>builder()
+            .fPut(MapperService.SINGLE_MAPPING_NAME, mappings).build();
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetaData.java]: it is not very clear, why we can't have `mappings` just as `CompressedXContent ` and have `this.mappings = mappings`?

**Review by mayya-sharipova — COMMENTED** [review]
Thanks @romseygeek, for now I would like to clarify some things

**Comment by romseygeek** [comment]
I've opened #52961 as a replacement here, which just changes the method signature of `getMappings()` and leaves the internal representations in place, which is a lot simpler.

**Pull request closed without merging**